### PR TITLE
[MIG] stock_reserve_sale: Migration to 12.0

### PR DESCRIPTION
**Pull request by hzh0292**

### CODE REVIEW

**Comment by pedrobaeza** [comment]
I'm afraid this still lacks the commit history. You can see it because there's only one commit.

**Comment by hzh0292** [comment]
> I'm afraid this still lacks the commit history. You can see it because there's only one commit.

So how can I do in the right way?

**Comment by pedrobaeza** [comment]
Following the given steps. Maybe you didn't replace where it says `$MODULE` with the proper module name.

**Comment by hzh0292** [comment]
> Following the given steps. Maybe you didn't replace where it says `$MODULE` with the proper module name.

I did fllowing the given steps and I already replaced the MODULE with the module name.

**Comment by pedrobaeza** [comment]
But the commits are not here... I'll take this branch later and try to fix it.

**Comment by rousseldenis** [comment]
@hzh0292 As the module is not present in 11, I suppose you took it from 10, so in git format patch, replace 11 by 10

**Comment by hzh0292** [comment]
> @hzh0292 As the module is not present in 11, I suppose you took it from 10, so in git format patch, replace 11 by 10

In fact,I took it from 9.0 branch.The 10.0 was not a complete migration from 8.0.

**Comment by rousseldenis** [comment]
> In fact,I took it from 9.0 branch.The 10.0 was not a complete migration from 8.0.

That looks better !

**Comment by rven** [comment]
Can you have a look at this https://github.com/OCA/stock-logistics-warehouse/issues/678?

**Comment by hzh0292** [comment]
> Can you have a look at this #678?

OK,I have seen this and I will check and add the code into the PR.

File: stock_reserve_sale/models/sale.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        self.release_all_stock_reservation()
+        return super(SaleOrder, self).action_cancel()
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    @api.multi
+    def _get_line_rule(self):
+        if self.route_id and self.route_id.rule_ids:
```
Comment (rven) [inline stock_reserve_sale/models/sale.py:63]: Add self.ensure_one()

File: stock_reserve_sale/models/sale.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    @api.multi
+    def _get_line_rule(self):
+        self.ensure_one()
+        if self.route_id and self.route_id.rule_ids:
+            return self.route_id.rule_ids.sorted(
+                key=lambda r: (r.route_sequence, r.sequence))[0]
```
Comment (rousseldenis) [inline stock_reserve_sale/models/sale.py]: Prefer using 'first()'
```suggestion
                from odoo.fields import first
                return first(self.route_id.rule_ids.sorted(
                key=lambda r: (r.route_sequence, r.sequence)))
```

**Review by rousseldenis — APPROVED** [review]
Code review. Minor comment.

Could you squash your commits into Migration one ?

**Comment by rousseldenis** [comment]
Depends on #637 

@hzh0292 Could you check travis ?

File: stock_reserve_sale/models/sale.py [hunk 3]
```diff
@@ -60,10 +60,9 @@ class SaleOrderLine(models.Model):
     def _get_line_rule(self):
         self.ensure_one()
         if self.route_id and self.route_id.rule_ids:
-            return self.route_id.rule_ids.sorted(
-                from odoo.fields import first
-                return first(self.route_id.rule_ids.sorted(
-                key=lambda r: (r.route_sequence, r.sequence)))
+            from odoo.fields import first
```
Comment (rousseldenis) [inline stock_reserve_sale/models/sale.py]: Put import on beginning of the file

**Comment by rousseldenis** [comment]
@hzh0292 Could you squash your commits and check tests?

**Comment by dreispt** [comment]
Migration commits need to be squashed.

**Comment by rousseldenis** [comment]
@hzh0292

**Pull request closed without merging**